constantine/math/arithmetic/assembly/limbs_asm_modular_arm64.nim

```diff
@@ -0,0 +1,208 @@
+  if spareBits >= 1:
+    # Addition cannot overflow u256, u384, ...
+    # v = u - M
+    for i in 0 ..< N:
+      if i == 0:
+        ctx.subs v[i], u[i], v[i]
+      else:
+        ctx.sbcs v[i], u[i], v[i]
+
+      # Next iteration
+      if i+2 < N:
+        ctx.ldr v[i+2], M[i+2]
+
+    # if carry clear u < M, so pick u
+    for i in 0 ..< N:
+      ctx.csel u[i], u[i], v[i], cc
+      ctx.str u[i], r[i]
+  else:
+    # Addition can overflow u256, u384, ...
+    let carryReg = b.reuseRegister()
+    ctx.adc carryReg, xzr, xzr
+
+    # v = u - M
+    for i in 0 ..< N:
+      if i == 0:
+        ctx.subs v[i], u[i], v[i]
+      else:
+        ctx.sbcs v[i], u[i], v[i]
+
+      # Next iteration
+      if i+2 < N:
+        ctx.ldr v[i+2], M[i+2]
+
+    # If it underflows here, it means that it was
+    # smaller than the modulus and we don't need `v`
+    ctx.sbcs xzr, carryReg, xzr
+
+    # if carry clear u < M, so pick u
+    for i in 0 ..< N:
+      ctx.csel u[i], u[i], v[i], cc
+      ctx.str u[i], r[i]
+
+  result.add ctx.generate()
+
+func addmod_asm*(r: var Limbs, a, b, M: Limbs, spareBits: static int) =
+  ## Constant-time modular addition
+  addmod_gen(r, a, b, M, spareBits)
+
+# Field subtraction
+# ------------------------------------------------------------
+
+macro submod_gen[N: static int](r_PIR: var Limbs[N], a_PIR, b_PIR, M_PIR: Limbs[N]): untyped =
+  ## Generate an optimized modular subtraction kernel
+  result = newStmtList()
+
+  var ctx = init(Assembler_arm64, BaseType)
+  let
+    # MemOffsettable is the better constraint but
+    # with ARM64 we cannot generate array offsets from it due to inline ASM auto-bracketings
+    r = asmArray(r_PIR, N, PointerInReg, asmInput, memIndirect = memWrite)
+    a = asmArray(a_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+    b = asmArray(b_PIR, N, PointerInReg, asmInputOutputEarlyClobber, memIndirect = memRead)
+    # We could force m as immediate by specializing per moduli
+    M = asmArray(M_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+
+    # Temp storage
+    uSym = ident"u"
+    vSym = ident"v"
+    u = asmArray(uSym, N, ElemsInReg, asmOutputEarlyClobber)
+    v = asmArray(vSym, N, ElemsInReg, asmOutputEarlyClobber)
+
+  # Prologue
+  result.add quote do:
+    var `usym`{.noinit, used.}, `vsym` {.noInit, used.}: typeof(`a_PIR`)
+
+  # Algorithm
+  if N >= 2:
+    ctx.ldp u[0], u[1], a[0]
+    ctx.ldp v[0], v[1], b[0]
+  else:
+    ctx.ldr u[0], a[0]
+    ctx.ldr v[0], b[0]
+
+  # u = a + b
+  for i in 0 ..< N:
+    if i == 0:
+      ctx.subs u[i], u[i], v[i]
+    else:
+      ctx.sbcs u[i], u[i], v[i]
+
+    # Next iteration
+    if i+2 < N:
+      ctx.ldr u[i+2], a[i+2]
+      ctx.ldr v[i+2], b[i+2]
+    elif 0 <= i-(N-2):
+      # Preload M
+      if i-(N-2) >= N:
+        # This can only occur if N == 1, for example in t_multilinear_extensions
+        ctx.ldr v[0], M[0]
+      else:
+        ctx.ldr v[i-(N-2)], M[i-(N-2)]
```

Comment:
For now this uses fancy prefetching but unsure if beneficial on Apple Silicon (can fetch/decode up to 8 instructions per cycle) or Raspberry Pi 5